compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs

```diff
@@ -520,68 +501,109 @@ pub fn report_dyn_incompatibility<'tcx>(
         }
     }
 
+    attempt_dyn_to_enum_suggestion(tcx, trait_def_id, &*trait_str, &mut err);
+
+    err
+}
+
+/// Attempt to suggest converting the `dyn Trait` argument to an enumeration
+/// over the types that implement `Trait`.
+fn attempt_dyn_to_enum_suggestion(
+    tcx: TyCtxt<'_>,
+    trait_def_id: DefId,
+    trait_str: &str,
+    err: &mut Diag<'_>,
+) {
     let impls_of = tcx.trait_impls_of(trait_def_id);
-    let impls = if impls_of.blanket_impls().is_empty() {
-        impls_of
-            .non_blanket_impls()
-            .values()
-            .flatten()
-            .filter(|def_id| {
-                !matches!(tcx.type_of(*def_id).instantiate_identity().kind(), ty::Dynamic(..))
-            })
-            .collect::<Vec<_>>()
-    } else {
-        vec![]
-    };
-    let externally_visible = if !impls.is_empty()
-        && let Some(def_id) = trait_def_id.as_local()
+
+    if !impls_of.blanket_impls().is_empty() {
+        return;
+    }
+
+    let concrete_impls: Option<Vec<Ty<'_>>> = impls_of
+        .non_blanket_impls()
+        .values()
+        .flatten()
+        .map(|impl_id| {
+            let Some(impl_type) = tcx.type_of(*impl_id).no_bound_vars() else { return None };
+            if let ty::Dynamic(..) = impl_type.kind() {
+                return None;
+            }
+            Some(impl_type)
+        })
+        .collect();
+    let Some(concrete_impls) = concrete_impls else { return };
+    if concrete_impls.is_empty() {
+        return;
+    }
```

Comment:
I did read https://github.com/rust-lang/rust/pull/133372#discussion_r1898091315. Returning early via the `collect::<Option<…>>()` does make sense albeit more so for the `no_bound_vars` case and less to for the `Dynamic` one,  I think? Not sure.

It means that we no longer provide the enum suggestion if there exists an "`impl for dyn`" whereas previously such self types were just filtered out. Doesn't matter much in practice I guess.

(I just realized that the `Dynamic` check seems to be a naive and incomplete attempt at filtering out obviously unsized types which would be invalid as enum payloads…)